File: frontend/src/types/codingQuestion.ts
export type CodingQuestionDifficulty = 'easy' | 'medium' | 'hard'

export interface CodingTestCase {
  id: number
  input: string
  expected_output: string
  is_sample: boolean
}

export interface CodingQuestion {
  id: number
  title: string
  description: string
  difficulty: CodingQuestionDifficulty
  job_id: number | null
  job_title: string | null
  starter_code: Record<string, string> | null
  created_at: string
  test_cases: CodingTestCase[]
}

export interface CodingTestCasePayload {
  input: string
  expected_output: string
  is_sample: boolean
}

export interface CodingQuestionPayload {
  title: string
  description: string
  difficulty: CodingQuestionDifficulty
  job_id: number | null
  starter_code: Record<string, string> | null
  test_cases: CodingTestCasePayload[]
}

File: app/schemas/coding_question.py
from datetime import datetime
from typing import Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field

from ..models import CodingQuestionDifficulty


class TestCasePayload(BaseModel):
    input: str
    expected_output: str
    is_sample: bool = False


class CodingQuestionCreate(BaseModel):
    title: str
    description: str
    difficulty: CodingQuestionDifficulty
    job_id: Optional[int] = None
    starter_code: Optional[Dict[str, str]] = None
    test_cases: List[TestCasePayload] = Field(default_factory=list)


class CodingQuestionUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    difficulty: Optional[CodingQuestionDifficulty] = None
    job_id: Optional[int] = None
    starter_code: Optional[Dict[str, str]] = None
    test_cases: Optional[List[TestCasePayload]] = None


class TestCaseOut(TestCasePayload):
    model_config = ConfigDict(from_attributes=True)

    id: int


class CodingQuestionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    description: str
    difficulty: CodingQuestionDifficulty
    job_id: Optional[int] = None
    job_title: Optional[str] = None
    starter_code: Optional[Dict[str, str]] = None
    created_at: datetime
    test_cases: List[TestCaseOut]


class CodingQuestionGenerateRequest(BaseModel):
    count: int = Field(default=2, ge=1, le=2)


class CodingQuestionGenerateResponse(BaseModel):
    questions: List["CodingQuestionGenerateItem"]


class CodingQuestionGenerateItem(BaseModel):
    title: str
    description: str
    difficulty: CodingQuestionDifficulty
    test_cases: List[TestCasePayload]
